plugins/gateway/tyk/v3.2.2/plugin.go

```diff
@@ -0,0 +1,257 @@
+func ResponseSendTelemetry(rw http.ResponseWriter, res *http.Response, req *http.Request) {
+	logger.Info("Handling telemetry")
+	telemetry, err := createTelemetry(res, req)
+	if err != nil {
+		logger.Errorf("Failed to create telemetry: %v", err)
+		return
+	}
+
+	apiClient := newAPIClient(host)
+	params := operations.NewPostTelemetryParams().WithBody(telemetry)
+
+	_, err = apiClient.Operations.PostTelemetry(params)
+	if err != nil {
+		logger.Errorf("Failed to post telemetry: %v", err)
+		return
+	}
+	logger.Infof("Telemetry has been sent")
+}
+
+func createTelemetry(res *http.Response, req *http.Request) (*models.Telemetry, error) {
+	truncatedBodyReq := false
+	truncatedBodyRes := false
+
+	apiDefinition := ctx.GetDefinition(req)
+	if apiDefinition == nil {
+		return nil, fmt.Errorf("failed to get api definition")
+	}
+
+	host, port := getHostAndPortFromTargetURL(apiDefinition.Proxy.TargetURL)
+	destinationNamespace := getDestinationNamespaceFromHost(host)
+
+	reqBody, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read request body: %v", err)
+	}
+	if len(reqBody) > MaxBodySize {
+		truncatedBodyReq = true
+		reqBody = []byte{}
+	}
+
+	resBody, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %v", err)
+	}
+	if len(resBody) > MaxBodySize {
+		truncatedBodyRes = true
+		resBody = []byte{}
+	}
+
+	pathAndQuery := getPathWithQuery(req.URL)
+
+	telemetry := models.Telemetry{
+		DestinationAddress:   ":" + port, // No destination ip for now
+		DestinationNamespace: destinationNamespace,
+		Request: &models.Request{
+			Common: &models.Common{
+				TruncatedBody: truncatedBodyReq,
+				Body:          strfmt.Base64(reqBody),
+				Headers:       createHeaders(req.Header),
+				Version:       req.Proto,
+			},
+			Host:   host,
+			Method: req.Method,
+			Path:   pathAndQuery,
+		},
+		RequestID: getRequestIDFromHeaders(req.Header),
+		Response: &models.Response{
+			Common: &models.Common{
+				TruncatedBody: truncatedBodyRes,
+				Body:          strfmt.Base64(resBody),
+				Headers:       createHeaders(res.Header),
+				Version:       res.Proto,
+			},
+			StatusCode: strconv.Itoa(res.StatusCode),
+		},
+		Scheme:        req.URL.Scheme,
+		SourceAddress: req.RemoteAddr,
+	}
+
+	return &telemetry, nil
+}
+
+func getRequestIDFromHeaders(reqHeaders http.Header) string {
+	if reqID, ok := reqHeaders[RequestIDHeaderKey]; ok {
+		return reqID[0]
+	}
+	return ""
+}
+
+func getPathWithQuery(reqURL *url.URL) string {
+	pathAndQuery := reqURL.Path
+	if reqURL.RawQuery != "" {
+		pathAndQuery += "?" + reqURL.RawQuery
+	}
+	return pathAndQuery
+}
+
+// Will try to extract the namespace from the host name, and if not found, will use the namespace that the gateway is running in.
+func getDestinationNamespaceFromHost(host string) string {
+	sp := strings.Split(host, ".")
+	if len(sp) >= MinimumSeparatedHostSize {
+		return sp[0]
+	}
+	return gatewayNamespace
+}
```

Comment:
need to add this support also in kong (or we always get the namespace there)?

Reply:
we always have the namespace there from what I saw